### PR TITLE
Fix multiplayer crash & Crafttweaker support enhancement & Fix #26

### LINKED ISSUE #26: Terrain Scanner leaves starting block empty, causing a "hole" in the bottom bedrock layer
This seems to happen whenever you generate a chunk directly via the Terrain Scanner. Flying below the bedrock, one will notice 1 block spaces in the bottom bedrock layer.

This does, however, not appear to occur if the chunk is pulled from the scanner queue instead of manually started.

### PR DESCRIPTION
**Pull request by misterplus**
Make this mod works in multiplayer
credits to one of my friends: https://github.com/Snownee

### CODE REVIEW

**Comment by Eladkay** [comment]
Looks good to me, would be grateful if someone could confirm

**Comment by FrostYang** [comment]
oh, misterplus saved my life!!!!
A few days ago, I saw this mod and add it to my server.Unfortunately，when I put the scanner down, my server crashed :(
I saw this PR and I tried it, my server just said "You are so smart !My dear Misterplus!!!"
And thank you for your mod too!!! It takes me a new world !

**Comment by misterplus** [comment]
@Eladkay I changed the crafttweaker syntax to support replacing specific blocks.
For example, using the old syntax, if I try to add diamond blocks, it'll also generate in the nether.
But now I can specify which block to replace.
Here's  the new syntax:
mods.scanner.Scanner.addRecipe(<minecraft:diamond_block>, <minecraft:stone>, 50, 0, 32);
(brackets are missing bc github is kinda weird)
This is kind of a useful change since it now kinda supports different dimensions.
But mainly I just wanted to use this mod in my upcoming modpack, the original syntax won't do the job for obvious reasons.

**Comment by misterplus** [comment]
Also renamed lang file to properly load it.

**Comment by misterplus** [comment]
Fixed #26

**Comment by misterplus** [comment]
Apprently the method getBiomeName is only cilent side, so I did some magic with access transformers and made sure that this mod actually works in multiplayer now.

If you need a jar file for confirming, pm me on discord:  Misterplus#0251

**Comment by misterplus** [comment]
Fixed TileEntities not getting copied correctly.
Also added a config option to prevent players from duping dungeon chests and blocks in general.

**Comment by Eladkay** [comment]
Hi, I will message you on Discord.

File: build.gradle [hunk 1]
```diff
@@ -11,7 +11,7 @@ apply plugin: 'net.minecraftforge.gradle.forge'
 //Only edit below this line, the above code adds and enables the necessary things for Forge to be setup.
 
 
-version = "1.6.3"
+version = "1.6.5"
```
Comment (Eladkay) [inline build.gradle]: version bump (why by 2?)

File: build.gradle [hunk 2]
```diff
@@ -97,4 +97,11 @@ processResources {
     from(sourceSets.main.resources.srcDirs) {
         exclude 'mcmod.info'
     }
+    rename '(.+_at.cfg)', 'META-INF/$1'
 }
+
+jar {
```
Comment (Eladkay) [inline build.gradle:103]: for ATs

File: src/main/resources/assets/scanner/lang/en_us.lang [hunk 3]
```diff
@@ -4,4 +4,4 @@ tile.biomeScannerAdv.name=Biome Scanner (Advanced)
 tile.biomeScannerElite.name=Biome Scanner (Elite)
 tile.biomeScannerUltimate.name=Biome Scanner (Ultimate)
 tile.scannerQueue.name=Scanner Queue
-itemGroup.scanner=Scanner
\ No newline at end of file
+itemGroup.scanner=Scanner
```
Comment (Eladkay) [inline src/main/resources/assets/scanner/lang/en_us.lang:7]: ok

File: src/main/resources/scanner_at.cfg [hunk 4]
```diff
@@ -0,0 +1 @@
+public net.minecraft.world.biome.Biome field_76791_y # biomeName
```
Comment (Eladkay) [inline src/main/resources/scanner_at.cfg:1]: https://media.discordapp.net/attachments/686506842221969440/687302234148438119/unknown.png

File: src/main/java/eladkay/scanner/terrain/TileEntityTerrainScanner.java [hunk 5]
```diff
@@ -72,14 +72,14 @@ public TileEntityTerrainScanner() {
...
-            current.setPos(getPos().getX() + 1, 0, getPos().getZ());
+            current.setPos(getPos().getX(), 0, getPos().getZ());
             changeState(true);
         }
     }
 
     public void activate() {
         changeState(true);
-        current.setPos(getPosStart().getX() + 1, 0, getPosStart().getZ());
+        current.setPos(getPosStart().getX(), 0, getPosStart().getZ());
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/terrain/TileEntityTerrainScanner.java:82]: the alleged fix for #26

File: src/main/java/eladkay/scanner/compat/ScannerCT.java [hunk 6]
```diff
@@ -36,9 +38,12 @@ public Add(ItemStack stack, int rarity, int minY, int maxY) {
         @Override
         public void apply() {
             if (!(stack.getItem() instanceof ItemBlock)) return;
+            if (!(materialStack.getItem() instanceof ItemBlock)) return;
             Block block = ((ItemBlock) stack.getItem()).getBlock();
             IBlockState blockState = block.getStateFromMeta(stack.getItemDamage());
-            Oregistry.registerEntry(new Oregistry.Entry(blockState, rarity, maxY, minY));
+            Block materialBlock = ((ItemBlock) materialStack.getItem()).getBlock();
+            IBlockState materialState = materialBlock.getStateFromMeta(materialStack.getItemDamage());
+            Oregistry.registerEntry(new Oregistry.Entry(blockState, materialState, rarity, maxY, minY));
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/compat/ScannerCT.java:46]: ok

File: src/main/java/eladkay/scanner/RegistryEventHandler.java [hunk 7]
```diff
@@ -23,6 +23,9 @@
     @SubscribeEvent
     public static void registerBlocks(RegistryEvent.Register<Block> event) {
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/RegistryEventHandler.java:24]: supposed fix for MP crash

File: src/main/java/eladkay/scanner/Config.java [hunk 8]
```diff
@@ -30,6 +31,7 @@ public static void initConfig(File configFile) {
         dimid = config.get("Scanner", "dimid", 99, "The ID for the fake overworld dimension").getInt();
         showOutline = config.get("Scanner", "showOutline", true, "Should the Terrain Scanner show its area of effect outline with shiny particles").getBoolean();
         genExtraVanillaOres = config.get("Scanner", "genVanillaOres", false, "Should the terrain scanner automatically generate extra vanilla ores? MineTweakered ores will always be spawned.").getBoolean();
+        voidOriginalBlock = config.get("Scanner", "voidOriginalBlock", false, "Should the terrain scanner void the original block after it copies the block?").getBoolean();
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/Config.java:34]: voids in the remote world, that is

File: src/main/java/eladkay/scanner/terrain/TileEntityTerrainScanner.java [hunk 9]
```diff
@@ -143,9 +143,8 @@ else if (getWorld().provider.getDimension() == 1)
             if ((local.getBlock().isReplaceable(getWorld(), imm) || local.getBlock().isAir(local, getWorld(), imm)) && !(local.getBlock() instanceof BlockFluidBase) && !(local.getBlock() instanceof BlockLiquid)) {
                 getWorld().setBlockState(imm, remote, 2);
                 if (remoteTE != null) {
-                    NBTTagCompound tag = new NBTTagCompound();
-                    remoteTE.writeToNBT(tag);
-                    getWorld().getTileEntity(imm).writeToNBT(tag);
+                    NBTTagCompound tag = remoteTE.serializeNBT();
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/terrain/TileEntityTerrainScanner.java:146]: fix for copying TEs

**Review by Eladkay — APPROVED** [review]
Excellent work. I had a chat with you on Discord and you explained the changes so after the quick fix I asked, I will merge it into a 1.12.2 branch

File: build.gradle [hunk 10]
```diff
@@ -11,7 +11,7 @@ apply plugin: 'net.minecraftforge.gradle.forge'
 //Only edit below this line, the above code adds and enables the necessary things for Forge to be setup.
 
 
-version = "1.6.3"
+version = "1.6.4"
```
Comment (Eladkay) [inline build.gradle:14]: ok

File: src/main/java/eladkay/scanner/ScannerMod.java [hunk 11]
```diff
@@ -53,7 +53,7 @@
     public static ScannerMod instance;
     public final static CreativeTabs TAB = new ScannerCreativeTabs(MODID);
     public static BlockAirey air;
-    public static final String VERSION = "1.6.3";
+    public static final String VERSION = "1.6.4";
```
Comment (Eladkay) [inline src/main/java/eladkay/scanner/ScannerMod.java:56]: ok

**Comment by Eladkay** [comment]
closed by #34, thank you :D

**Pull request closed without merging**